src/main/java/com/glencoesoftware/omero/ms/image/region/ShapeMaskRequestHandler.java

```diff
@@ -291,42 +291,218 @@ public ShapeMaskRequestHandler(ShapeMaskCtx shapeMaskCtx, String labelImagePath,
+            }
+            case TILEDB_UINT64:
+            case TILEDB_INT64: {
+                long[] subarrayDomain = new long[(int) num_dims*2];
+                for(int i = 0; i < num_dims; i++) {
+                    long start = (long) (domain.getDimension(i).getDomain().getFirst());
+                    long end = (long) domain.getDimension(i).getDomain().getSecond();
+                    subarrayDomain[i*2] = start;
+                    subarrayDomain[i*2 + 1] = end;
+                }
+                return new NativeArray(ctx, subarrayDomain, dimType);
+            }
+            default: return null;
+        }
+    }
+
     private byte[] getData(Array array, Context ctx) throws TileDBError {
         ArraySchema schema = array.getSchema();
         Domain domain = schema.getDomain();
-        long num_dims = domain.getNDim();
         Attribute attribute = schema.getAttribute("a1");
-        int[] subarrayDomain = new int[(int) num_dims*2];
-        int capacity = 1;
-        for(int i = 0; i < num_dims; i++) {
-            int start = (int) (domain.getDimension(i).getDomain().getFirst());
-            int end = (int) domain.getDimension(i).getDomain().getSecond();
-            subarrayDomain[i*2] = start;
-            subarrayDomain[i*2 + 1] = end;
-            capacity *= (end - start + 1);
-        }
+
         int bytesPerPixel = 1;
-        if (attribute.getType() == Datatype.TILEDB_UINT32 || attribute.getType() == Datatype.TILEDB_INT32) {
-            bytesPerPixel = 4;
+        switch (attribute.getType()) {
+            case TILEDB_UINT8:
+            case TILEDB_INT8:
+                break;
+            case TILEDB_UINT16:
+            case TILEDB_INT16:
+                bytesPerPixel = 2;
+                break;
+            case TILEDB_UINT32:
+            case TILEDB_INT32:
+                bytesPerPixel = 4;
+                break;
+            case TILEDB_UINT64:
+            case TILEDB_INT64:
+                bytesPerPixel = 8;
+                break;
+            default:
+                throw new IllegalArgumentException("Attribute type " + attribute.getType().toString() + " not supported");
         }
-        capacity *= bytesPerPixel;
+        int capacity = getCapacity(domain) * bytesPerPixel;
         ByteBuffer buffer = ByteBuffer.allocateDirect(capacity);
         ByteBuffer bufferSlice = buffer.slice();
         bufferSlice.order(ByteOrder.nativeOrder());
-        NativeArray subArray = new NativeArray(ctx, subarrayDomain, Integer.class);
+        //Dimensions in Dense Arrays must be the same type
         Query query = new Query(array, QueryType.TILEDB_READ);
-        query.setSubarray(subArray);
+        query.setSubarray(getSubarray(ctx, domain));
         query.setBuffer("a1", bufferSlice);
         query.submit();
-        byte[] outputBytes = new byte[buffer.capacity()/bytesPerPixel];
+        byte[] outputBytes = new byte[capacity];
+        buffer.get(outputBytes);
+        query.close();
```

Comment:
We should not be closing the query explicitly but rather always using a try with resources block.